Quadriga/src/main/java/edu/asu/spring/quadriga/web/workspace/AddTextController.java

```diff
@@ -0,0 +1,138 @@
+     *            Generic HTTP Response Object
+     * @param txtFile
+     *            Web backing domain object to store the Text Details
+     * @param result
+     *            Generic BingingResults object to hold validation errors
+     * @param workspaceid
+     *            Path Variable to retrieve the associated Workspace id
+     * @return Returns the workspace details page if success else returns the
+     *         add text page.
+     * @throws QuadrigaStorageException
+     * @throws QuadrigaAccessException
+     * @throws IOException
+     */
+    @AccessPolicies({
+            @ElementAccessPolicy(type = CheckedElementType.WORKSPACE, paramIndex = 4, userRole = {
+                    RoleNames.ROLE_WORKSPACE_COLLABORATOR_ADMIN, RoleNames.ROLE_WORKSPACE_COLLABORATOR_EDITOR }),
+            @ElementAccessPolicy(type = CheckedElementType.PROJECT, paramIndex = 5, userRole = {
+                    RoleNames.ROLE_COLLABORATOR_ADMIN, RoleNames.ROLE_PROJ_COLLABORATOR_ADMIN,
+                    RoleNames.ROLE_PROJ_COLLABORATOR_CONTRIBUTOR }) })
+    @RequestMapping(value = "/auth/workbench/workspace/{projectid}/{workspaceid}/addtext", method = RequestMethod.POST)
+    public ModelAndView saveTextFileForm(HttpServletResponse resp,
+            @Validated @ModelAttribute("textfile") TextFile txtFile, BindingResult result,
+            @PathVariable("workspaceid") String workspaceid, @PathVariable("projectid") String projid,
+            RedirectAttributes redirectAttributes)
+                    throws QuadrigaAccessException, QuadrigaStorageException {
+
+        ModelAndView model = new ModelAndView();
+        model.getModelMap().put("workspaceId", workspaceid);
+        model.getModelMap().put("myProjectId", projid);
+        if (result.hasErrors()) {
+            model.setViewName("auth/workbench/workspace/addtext");
+            model.getModelMap().put("textfile", txtFile);
+        } else {
+            model = new ModelAndView("redirect:/auth/workbench/workspace/workspacedetails/" + workspaceid);
+            txtFile.setWorkspaceId(workspaceid);
+            txtFile.setProjectId(projid);
+            try {
+                if (tfManager.saveTextFile(txtFile)) {
+                    tfManager.retrieveTextFiles(workspaceid);
+                    redirectAttributes.addFlashAttribute("show_success_alert", true);
+                    redirectAttributes.addFlashAttribute("success_alert_msg", "The text file is successfully saved");
+                } else {
+                    redirectAttributes.addFlashAttribute("show_error_alert", true);
+                    redirectAttributes.addFlashAttribute("error_alert_msg", "Error while saving the text file.");
+                }
+            } catch (FileStorageException fse) {
+                redirectAttributes.addFlashAttribute("show_error_alert", true);
+                redirectAttributes.addFlashAttribute("error_alert_msg", "Unable to save text file. Please check the default directory.");
```

Comment:
Should say: "The file could not be saved. If the problem persists, please contact an administrator."